// @nestjs/core/injector/instance-loader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const injector_1 = require("./injector");
const common_1 = require("@nestjs/common");
const messages_1 = require("../helpers/messages");
class InstanceLoader {
    constructor(container) {
        this.container = container;
        this.injector = new injector_1.Injector();
        this.logger = new common_1.Logger(InstanceLoader.name);
    }
    createInstancesOfDependencies() {
        const modules = this.container.getModules();
        this.createPrototypes(modules);
        this.createInstances(modules);
    }
    createPrototypes(modules) {
        modules.forEach((module) => {
            this.createPrototypesOfComponents(module);
            this.createPrototypesOfRoutes(module);
        });
    }
    createInstances(modules) {
        modules.forEach((module) => {
            this.createInstancesOfComponents(module);
            this.createInstancesOfRoutes(module);
            const { name } = module.metatype;
            this.logger.log(messages_1.ModuleInitMessage(name));
        });
    }
    createPrototypesOfComponents(module) {
        module.components.forEach((wrapper) => {
            this.injector.loadPrototypeOfInstance(wrapper, module.components);
        });
    }
    createInstancesOfComponents(module) {
        module.components.forEach((wrapper) => {
            this.injector.loadInstanceOfComponent(wrapper, module);
        });
    }
    createPrototypesOfRoutes(module) {
        module.routes.forEach((wrapper) => {
            this.injector.loadPrototypeOfInstance(wrapper, module.routes);
        });
    }
    createInstancesOfRoutes(module) {
        module.routes.forEach((wrapper) => {
            this.injector.loadInstanceOfRoute(wrapper, module);
        });
    }
}
exports.InstanceLoader = InstanceLoader;
